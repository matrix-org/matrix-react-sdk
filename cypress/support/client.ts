/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/// <reference types="cypress" />

import type {
    MatrixClient,
    Room,
    IContent,
    FileType,
    Upload,
    UploadOpts,
    ICreateRoomOpts,
    ISendEventResponse,
} from "matrix-js-sdk/src/matrix";
import Chainable = Cypress.Chainable;

declare global {
    // eslint-disable-next-line @typescript-eslint/no-namespace
    namespace Cypress {
        interface Chainable {
            /**
             * Returns the MatrixClient from the MatrixClientPeg
             */
            getClient(): Chainable<MatrixClient | undefined>;
            /**
             * Create a room with given options.
             * @param options the options to apply when creating the room
             * @return the ID of the newly created room
             */
            createRoom(options: ICreateRoomOpts): Chainable<string>;
            /**
             * Create a space with given options.
             * @param options the options to apply when creating the space
             * @return the ID of the newly created space (room)
             */
            createSpace(options: ICreateRoomOpts): Chainable<string>;
            /**
             * Invites the given user to the given room.
             * @param roomId the id of the room to invite to
             * @param userId the id of the user to invite
             */
            inviteUser(roomId: string, userId: string): Chainable<{}>;
            /**
             * Sets account data for the user.
             * @param type The type of account data.
             * @param data The data to store.
             */
            setAccountData(type: string, data: object): Chainable<{}>;
            /**
             * @param {string} roomId
             * @param {string} threadId
             * @param {string} eventType
             * @param {Object} content
             * @return {module:http-api.MatrixError} Rejects: with an error response.
             */
            sendEvent(
                roomId: string,
                threadId: string | null,
                eventType: string,
                content: IContent,
            ): Chainable<ISendEventResponse>;
            /**
             * @param {string} name
             * @param {module:client.callback} callback Optional.
             * @return {Promise} Resolves: {} an empty object.
             * @return {module:http-api.MatrixError} Rejects: with an error response.
             */
            setDisplayName(name: string): Chainable<{}>;
            /**
             * @param {string} url
             * @param {module:client.callback} callback Optional.
             * @return {Promise} Resolves: {} an empty object.
             * @return {module:http-api.MatrixError} Rejects: with an error response.
             */
            setAvatarUrl(url: string): Chainable<{}>;
            /**
             * Upload a file to the media repository on the homeserver.
             *
             * @param {object} file The object to upload. On a browser, something that
             *   can be sent to XMLHttpRequest.send (typically a File).  Under node.js,
             *   a a Buffer, String or ReadStream.
             */
            uploadContent(file: FileType, opts?: UploadOpts): Chainable<Awaited<Upload["promise"]>>;
            /**
             * Turn an MXC URL into an HTTP one. <strong>This method is experimental and
             * may change.</strong>
             * @param {string} mxcUrl The MXC URL
             * @param {Number} width The desired width of the thumbnail.
             * @param {Number} height The desired height of the thumbnail.
             * @param {string} resizeMethod The thumbnail resize method to use, either
             * "crop" or "scale".
             * @param {Boolean} allowDirectLinks If true, return any non-mxc URLs
             * directly. Fetching such URLs will leak information about the user to
             * anyone they share a room with. If false, will return null for such URLs.
             * @return {?string} the avatar URL or null.
             */
            mxcUrlToHttp(
                mxcUrl: string,
                width?: number,
                height?: number,
                resizeMethod?: string,
                allowDirectLinks?: boolean,
            ): string | null;
            /**
             * Gets the list of DMs with a given user
             * @param userId The ID of the user
             * @return the list of DMs with that user
             */
            getDmRooms(userId: string): Chainable<string[]>;
            /**
             * Joins the given room by alias or ID
             * @param roomIdOrAlias the id or alias of the room to join
             */
            joinRoom(roomIdOrAlias: string): Chainable<Room>;
        }
    }
}

Cypress.Commands.add("getClient", (): Chainable<MatrixClient | undefined> => {
    return cy.window({ log: false }).then((win) => win.mxMatrixClientPeg.matrixClient);
});

Cypress.Commands.add("getDmRooms", (userId: string): Chainable<string[]> => {
    return cy
        .getClient()
        .then((cli) => cli.getAccountData("m.direct")?.getContent<Record<string, string[]>>())
        .then((dmRoomMap) => dmRoomMap[userId] ?? []);
});

Cypress.Commands.add("createRoom", (options: ICreateRoomOpts): Chainable<string> => {
    return cy.window({ log: false }).then(async (win) => {
        const cli = win.mxMatrixClientPeg.matrixClient;
        const resp = await cli.createRoom(options);
        const roomId = resp.room_id;

        if (!cli.getRoom(roomId)) {
            await new Promise<void>((resolve) => {
                const onRoom = (room: Room) => {
                    if (room.roomId === roomId) {
                        cli.off(win.matrixcs.ClientEvent.Room, onRoom);
                        resolve();
                    }
                };
                cli.on(win.matrixcs.ClientEvent.Room, onRoom);
            });
        }

        return roomId;
    });
});

Cypress.Commands.add("createSpace", (options: ICreateRoomOpts): Chainable<string> => {
    return cy.createRoom({
        ...options,
        creation_content: {
            type: "m.space",
        },
    });
});

Cypress.Commands.add("inviteUser", (roomId: string, userId: string): Chainable<{}> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        const res = await cli.invite(roomId, userId);
        Cypress.log({ name: "inviteUser", message: `sent invite in ${roomId} for ${userId}` });
        return res;
    });
});

Cypress.Commands.add("setAccountData", (type: string, data: object): Chainable<{}> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        return cli.setAccountData(type, data);
    });
});

Cypress.Commands.add(
    "sendEvent",
    (roomId: string, threadId: string | null, eventType: string, content: IContent): Chainable<ISendEventResponse> => {
        return cy.getClient().then(async (cli: MatrixClient) => {
            return cli.sendEvent(roomId, threadId, eventType, content);
        });
    },
);

Cypress.Commands.add("setDisplayName", (name: string): Chainable<{}> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        return cli.setDisplayName(name);
    });
});

Cypress.Commands.add("uploadContent", (file: FileType, opts?: UploadOpts): Chainable<Awaited<Upload["promise"]>> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        return cli.uploadContent(file, opts);
    });
});

Cypress.Commands.add("setAvatarUrl", (url: string): Chainable<{}> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        return cli.setAvatarUrl(url);
    });
});

Cypress.Commands.add("joinRoom", (roomIdOrAlias: string): Chainable<Room> => {
    return cy.getClient().then((cli) => cli.joinRoom(roomIdOrAlias));
});
