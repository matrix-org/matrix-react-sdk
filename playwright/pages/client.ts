/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { JSHandle, Page } from "@playwright/test";
import { PageFunctionOn } from "playwright-core/types/structs";

import type {
    IContent,
    ICreateRoomOpts,
    ISendEventResponse,
    MatrixClient,
    Room,
    IRoomDirectoryOptions,
} from "matrix-js-sdk/src/matrix";

export class Client {
    protected client: JSHandle<MatrixClient>;

    protected getClientHandle(): Promise<JSHandle<MatrixClient>> {
        return this.page.evaluateHandle(() => window.mxMatrixClientPeg.get());
    }

    public async prepareClient(): Promise<JSHandle<MatrixClient>> {
        if (!this.client) {
            this.client = await this.getClientHandle();
        }
        return this.client;
    }

    public constructor(protected readonly page: Page) {
        page.on("framenavigated", async () => {
            this.client = null;
        });
    }

    public evaluate<R, Arg, O extends MatrixClient = MatrixClient>(
        pageFunction: PageFunctionOn<O, Arg, R>,
        arg: Arg,
    ): Promise<R>;
    public evaluate<R, O extends MatrixClient = MatrixClient>(
        pageFunction: PageFunctionOn<O, void, R>,
        arg?: any,
    ): Promise<R>;
    public async evaluate<T>(fn: (client: MatrixClient) => T, arg?: any): Promise<T> {
        await this.prepareClient();
        return this.client.evaluate(fn, arg);
    }

    /**
     * @param roomId ID of the room to send the event into
     * @param threadId ID of the thread to send into or null for main timeline
     * @param eventType type of event to send
     * @param content the event content to send
     */
    public async sendEvent(
        roomId: string,
        threadId: string | null,
        eventType: string,
        content: IContent,
    ): Promise<ISendEventResponse> {
        const client = await this.prepareClient();
        return client.evaluate(
            async (client, { roomId, threadId, eventType, content }) => {
                return client.sendEvent(roomId, threadId, eventType, content);
            },
            { roomId, threadId, eventType, content },
        );
    }

    /**
     * Send a message as a bot into a room
     * @param roomId ID of the room to send the message into
     * @param content the event content to send
     */
    public async sendMessage(roomId: string, content: IContent): Promise<ISendEventResponse> {
        const client = await this.prepareClient();
        return client.evaluate(
            (client, { roomId, content }) => {
                return client.sendMessage(roomId, content);
            },
            {
                roomId,
                content,
            },
        );
    }

    /**
     * Create a room with given options.
     * @param options the options to apply when creating the room
     * @return the ID of the newly created room
     */
    public async createRoom(options: ICreateRoomOpts): Promise<string> {
        const client = await this.prepareClient();
        return await client.evaluate(async (cli, options) => {
            const resp = await cli.createRoom(options);
            const roomId = resp.room_id;
            if (!cli.getRoom(roomId)) {
                await new Promise<void>((resolve) => {
                    const onRoom = (room: Room) => {
                        if (room.roomId === roomId) {
                            cli.off(window.matrixcs.ClientEvent.Room, onRoom);
                            resolve();
                        }
                    };
                    cli.on(window.matrixcs.ClientEvent.Room, onRoom);
                });
            }
            return roomId;
        }, options);
    }

    /**
     * Create a space with given options.
     * @param options the options to apply when creating the space
     * @return the ID of the newly created space (room)
     */
    public async createSpace(options: ICreateRoomOpts): Promise<string> {
        return this.createRoom({
            ...options,
            creation_content: {
                ...options.creation_content,
                type: "m.space",
            },
        });
    }

    /**
     * Joins the given room by alias or ID
     * @param roomIdOrAlias the id or alias of the room to join
     */
    public async joinRoom(roomIdOrAlias: string): Promise<void> {
        const client = await this.prepareClient();
        await client.evaluate(async (client, roomIdOrAlias) => {
            return await client.joinRoom(roomIdOrAlias);
        }, roomIdOrAlias);
    }

    /**
     * Make this bot join a room by name
     * @param roomName Name of the room to join
     */
    public async joinRoomByName(roomName: string): Promise<void> {
        const client = await this.prepareClient();
        await client.evaluate(
            (client, { roomName }) => {
                const room = client.getRooms().find((r) => r.getDefaultRoomName(client.getUserId()) === roomName);
                if (room) {
                    return client.joinRoom(room.roomId);
                }
                throw new Error(`Bot room join failed. Cannot find room '${roomName}'`);
            },
            {
                roomName,
            },
        );
    }

    /**
     * Invites the given user to the given room.
     * @param roomId the id of the room to invite to
     * @param userId the id of the user to invite
     */
    public async inviteUser(roomId: string, userId: string): Promise<void> {
        const client = await this.prepareClient();
        await client.evaluate((client, { roomId, userId }) => client.invite(roomId, userId), {
            roomId,
            userId,
        });
    }

    public async publicRooms(options?: IRoomDirectoryOptions): ReturnType<MatrixClient["publicRooms"]> {
        const client = await this.prepareClient();
        return await client.evaluate((client, options) => {
            return client.publicRooms(options);
        }, options);
    }
}
