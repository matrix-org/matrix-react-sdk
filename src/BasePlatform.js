// @flow

/*
Copyright 2016 Aviral Dasgupta
Copyright 2016 OpenMarket Ltd
Copyright 2018 New Vector Ltd
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import {MatrixClient} from "matrix-js-sdk";
import dis from './dispatcher';
import BaseEventIndexManager from './indexing/BaseEventIndexManager';

/**
 * Base class for classes that provide platform-specific functionality
 * eg. Setting an application badge or displaying notifications
 *
 * Instances of this class are provided by the application.
 */
export default class BasePlatform {
    constructor() {
        this.notificationCount = 0;
        this.errorDidOccur = false;

        dis.register(this._onAction.bind(this));
    }

    _onAction(payload: Object) {
        switch (payload.action) {
            case 'on_client_not_viable':
            case 'on_logged_out':
                this.setNotificationCount(0);
                break;
        }
    }

    // Used primarily for Analytics
    getHumanReadableName(): string {
        return 'Base Platform';
    }

    setNotificationCount(count: number) {
        this.notificationCount = count;
    }

    setErrorStatus(errorDidOccur: boolean) {
        this.errorDidOccur = errorDidOccur;
    }

    /**
     * Returns true if the platform supports displaying
     * notifications, otherwise false.
     * @returns {boolean} whether the platform supports displaying notifications
     */
    supportsNotifications(): boolean {
        return false;
    }

    /**
     * Returns true if the application currently has permission
     * to display notifications. Otherwise false.
     * @returns {boolean} whether the application has permission to display notifications
     */
    maySendNotifications(): boolean {
        return false;
    }

    /**
     * Requests permission to send notifications. Returns
     * a promise that is resolved when the user has responded
     * to the request. The promise has a single string argument
     * that is 'granted' if the user allowed the request or
     * 'denied' otherwise.
     */
    requestNotificationPermission(): Promise<string> {
    }

    displayNotification(title: string, msg: string, avatarUrl: string, room: Object) {
    }

    loudNotification(ev: Event, room: Object) {
    }

    /**
     * Returns a promise that resolves to a string representing
     * the current version of the application.
     */
    getAppVersion(): Promise<string> {
        throw new Error("getAppVersion not implemented!");
    }

    /*
     * If it's not expected that capturing the screen will work
     * with getUserMedia, return a string explaining why not.
     * Otherwise, return null.
     */
    screenCaptureErrorString(): string {
        return "Not implemented";
    }

    /**
     * Restarts the application, without neccessarily reloading
     * any application code
     */
    reload() {
        throw new Error("reload not implemented!");
    }

    supportsAutoLaunch(): boolean {
        return false;
    }

    // XXX: Surely this should be a setting like any other?
    async getAutoLaunchEnabled(): boolean {
        return false;
    }

    async setAutoLaunchEnabled(enabled: boolean): void {
        throw new Error("Unimplemented");
    }

    supportsAutoHideMenuBar(): boolean {
        return false;
    }

    async getAutoHideMenuBarEnabled(): boolean {
        return false;
    }

    async setAutoHideMenuBarEnabled(enabled: boolean): void {
        throw new Error("Unimplemented");
    }

    supportsMinimizeToTray(): boolean {
        return false;
    }

    async getMinimizeToTrayEnabled(): boolean {
        return false;
    }

    async setMinimizeToTrayEnabled(enabled: boolean): void {
        throw new Error("Unimplemented");
    }

    /**
     * Get our platform specific EventIndexManager.
     *
     * @return {BaseEventIndexManager} The EventIndex manager for our platform,
     * can be null if the platform doesn't support event indexing.
     */
    getEventIndexingManager(): BaseEventIndexManager | null {
        return null;
    }

    setLanguage(preferredLangs: string[]) {}

    getSSOCallbackUrl(hsUrl: string, isUrl: string): URL {
        const url = new URL(window.location.href);
        // XXX: at this point, the fragment will always be #/login, which is no
        // use to anyone. Ideally, we would get the intended fragment from
        // MatrixChat.screenAfterLogin so that you could follow #/room links etc
        // through an SSO login.
        url.hash = "";
        url.searchParams.set("homeserver", hsUrl);
        url.searchParams.set("identityServer", isUrl);
        return url;
    }

    /**
     * Begin Single Sign On flows.
     * @param {MatrixClient} mxClient the matrix client using which we should start the flow
     * @param {"sso"|"cas"} loginType the type of SSO it is, CAS/SSO.
     */
    startSingleSignOn(mxClient: MatrixClient, loginType: "sso"|"cas") {
        const callbackUrl = this.getSSOCallbackUrl(mxClient.getHomeserverUrl(), mxClient.getIdentityServerUrl());
        window.location.href = mxClient.getSsoLoginUrl(callbackUrl.toString(), loginType); // redirect to SSO
    }

    onKeyDown(ev: KeyboardEvent): boolean {
        return false; // no shortcuts implemented
    }
}
