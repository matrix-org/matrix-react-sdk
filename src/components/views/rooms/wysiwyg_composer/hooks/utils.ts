/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { MutableRefObject } from "react";

import { TimelineRenderingType } from "../../../../../contexts/RoomContext";
import { IRoomState } from "../../../../structures/RoomView";

export function focusComposer(
    composerElement: MutableRefObject<HTMLElement | null>,
    renderingType: TimelineRenderingType,
    roomContext: IRoomState,
    timeoutId: MutableRefObject<number | null>,
): void {
    if (renderingType === roomContext.timelineRenderingType) {
        // Immediately set the focus, so if you start typing it
        // will appear in the composer
        composerElement.current?.focus();
        // If we call focus immediate, the focus _is_ in the right
        // place, but the cursor is invisible, presumably because
        // some other event is still processing.
        // The following line ensures that the cursor is actually
        // visible in composer.
        if (timeoutId.current) {
            clearTimeout(timeoutId.current);
        }
        timeoutId.current = window.setTimeout(() => composerElement.current?.focus(), 200);
    }
}

export function setCursorPositionAtTheEnd(element: HTMLElement): void {
    const range = document.createRange();
    range.selectNodeContents(element);
    range.collapse(false);
    const selection = document.getSelection();
    selection.removeAllRanges();
    selection.addRange(range);

    element.focus();
}
