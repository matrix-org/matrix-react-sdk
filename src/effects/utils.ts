/*
 Copyright 2020 Nurjin Jafar
 Copyright 2020 Nordeck IT + Consulting GmbH.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
/**
 * Checks a message if it contains one of the provided emojis
 * @param  {Object} content The message
 * @param  {Array<string>} emojis The list of emojis to check for
 */
export const containsEmoji = (content: { msgtype: string, body: string }, emojis: Array<string>): boolean => {
    return emojis.some((emoji) => content.body && content.body.includes(emoji));
}
