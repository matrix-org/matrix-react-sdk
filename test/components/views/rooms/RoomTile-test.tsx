/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
// eslint-disable-next-line deprecate/import
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import { mocked } from "jest-mock";
import { MatrixClient } from "matrix-js-sdk/src/client";
import { Room } from "matrix-js-sdk/src/models/room";
import { RoomMember } from "matrix-js-sdk/src/models/room-member";

import {
    stubClient,
    mockStateEventImplementation,
    mkRoom,
    mkVideoChannelMember,
    stubVideoChannelStore,
    StubVideoChannelStore,
} from "../../../test-utils";
import { STUCK_DEVICE_TIMEOUT_MS } from "../../../../src/utils/VideoChannelUtils";
import RoomTile from "../../../../src/components/views/rooms/RoomTile";
import SettingsStore from "../../../../src/settings/SettingsStore";
import { DefaultTagID } from "../../../../src/stores/room-list/models";
import DMRoomMap from "../../../../src/utils/DMRoomMap";
import { MatrixClientPeg } from "../../../../src/MatrixClientPeg";
import PlatformPeg from "../../../../src/PlatformPeg";
import BasePlatform from "../../../../src/BasePlatform";

const mockGetMember = (room: Room, getMembership: (userId: string) => string = () => "join") => {
    mocked(room).getMember.mockImplementation(userId => ({
        userId,
        membership: getMembership(userId),
        name: userId,
        rawDisplayName: userId,
        roomId: "!1:example.org",
        getAvatarUrl: () => {},
        getMxcAvatarUrl: () => {},
    }) as unknown as RoomMember);
};

describe("RoomTile", () => {
    jest.spyOn(PlatformPeg, 'get')
        .mockReturnValue({ overrideBrowserShortcuts: () => false } as unknown as BasePlatform);

    let cli: MatrixClient;
    let store: StubVideoChannelStore;
    beforeEach(() => {
        const realGetValue = SettingsStore.getValue;
        SettingsStore.getValue = <T, >(name: string, roomId?: string): T => {
            if (name === "feature_video_rooms") {
                return true as unknown as T;
            }
            return realGetValue(name, roomId);
        };

        stubClient();
        cli = MatrixClientPeg.get();
        store = stubVideoChannelStore();
        DMRoomMap.makeShared();
    });

    afterEach(() => {
        jest.clearAllMocks();
        jest.useRealTimers();
    });

    describe("video rooms", () => {
        let room: Room;
        beforeEach(() => {
            room = mkRoom(cli, "!1:example.org");
            mocked(room.isElementVideoRoom).mockReturnValue(true);
        });

        const mountTile = () => mount(
            <RoomTile
                room={room}
                showMessagePreview={false}
                isMinimized={false}
                tag={DefaultTagID.Untagged}
            />,
        );

        it("tracks connection state", () => {
            const tile = mountTile();
            expect(tile.find(".mx_VideoRoomSummary_indicator").text()).toEqual("Video");

            act(() => { store.startConnect("!1:example.org"); });
            tile.update();
            expect(tile.find(".mx_VideoRoomSummary_indicator").text()).toEqual("Joiningâ€¦");

            act(() => { store.connect("!1:example.org"); });
            tile.update();
            expect(tile.find(".mx_VideoRoomSummary_indicator").text()).toEqual("Joined");

            act(() => { store.disconnect(); });
            tile.update();
            expect(tile.find(".mx_VideoRoomSummary_indicator").text()).toEqual("Video");
        });

        it("displays connected members", () => {
            mockGetMember(room, userId => userId === "@chris:example.org" ? "leave" : "join");
            mocked(room.currentState).getStateEvents.mockImplementation(mockStateEventImplementation([
                // A user connected from 2 devices
                mkVideoChannelMember("@alice:example.org", ["device 1", "device 2"]),
                // A disconnected user
                mkVideoChannelMember("@bob:example.org", []),
                // A user that claims to have a connected device, but has left the room
                mkVideoChannelMember("@chris:example.org", ["device 1"]),
            ]));

            const tile = mountTile();

            // Only Alice should display as connected
            expect(tile.find(".mx_VideoRoomSummary_participants").text()).toEqual("1");
        });

        it("reflects local echo in connected members", () => {
            mockGetMember(room);
            mocked(room.currentState).getStateEvents.mockImplementation(mockStateEventImplementation([
                // Make the remote echo claim that we're connected, while leaving the store disconnected
                mkVideoChannelMember(cli.getUserId(), [cli.getDeviceId()]),
            ]));

            const tile = mountTile();

            // Because of our local echo, we should still appear as disconnected
            expect(tile.find(".mx_VideoRoomSummary_participants").exists()).toEqual(false);
        });

        it("doesn't count members whose device data has expired", () => {
            jest.useFakeTimers();
            jest.setSystemTime(0);

            mockGetMember(room);
            mocked(room.currentState).getStateEvents.mockImplementation(mockStateEventImplementation([
                mkVideoChannelMember("@alice:example.org", ["device 1"], STUCK_DEVICE_TIMEOUT_MS),
            ]));

            const tile = mountTile();

            expect(tile.find(".mx_VideoRoomSummary_participants").text()).toEqual("1");
            // Expire Alice's device data
            act(() => { jest.advanceTimersByTime(STUCK_DEVICE_TIMEOUT_MS); });
            tile.update();
            expect(tile.find(".mx_VideoRoomSummary_participants").exists()).toEqual(false);
        });
    });
});
