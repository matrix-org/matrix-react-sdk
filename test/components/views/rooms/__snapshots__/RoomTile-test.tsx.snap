// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomTile when message previews are enabled and there is a message in a thread should render as expected 1`] = `
<DocumentFragment>
  <div
    aria-describedby="mx_RoomTile_messagePreview_!1:example.org"
    aria-label="!1:​example.org"
    aria-selected="false"
    class="mx_AccessibleButton mx_RoomTile"
    role="treeitem"
    tabindex="-1"
  >
    <div
      class="mx_DecoratedRoomAvatar"
    >
      <span
        class="mx_BaseAvatar"
        role="presentation"
      >
        <span
          aria-hidden="true"
          class="mx_BaseAvatar_initial"
          style="font-size: 20.8px; width: 32px; line-height: 32px;"
        >
          !
        </span>
        <img
          alt=""
          aria-hidden="true"
          class="mx_BaseAvatar_image"
          data-testid="avatar-img"
          loading="lazy"
          src="data:image/png;base64,00"
          style="width: 32px; height: 32px;"
        />
      </span>
    </div>
    <div
      class="mx_RoomTile_titleContainer"
    >
      <div
        class="mx_RoomTile_title mx_RoomTile_titleWithSubtitle"
        tabindex="-1"
        title="!1:​example.org"
      >
        <span
          dir="auto"
        >
          !1:​example.org
        </span>
      </div>
      <div
        class="mx_RoomTile_subtitle"
        id="mx_RoomTile_messagePreview_!1:example.org"
        title="test thread reply"
      >
        <span
          class="mx_RoomTile_subtitle_text"
        >
          test thread reply
        </span>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="mx_RoomTile_badgeContainer"
    />
    <div
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Room options"
      class="mx_AccessibleButton mx_RoomTile_menuButton"
      role="button"
      tabindex="0"
    />
    <div
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Notification options"
      class="mx_AccessibleButton mx_RoomTile_notificationsButton"
      role="button"
      tabindex="-1"
    />
  </div>
</DocumentFragment>
`;

exports[`RoomTile when message previews are enabled and there is a message in the room should render as expected 1`] = `
<DocumentFragment>
  <div
    aria-describedby="mx_RoomTile_messagePreview_!1:example.org"
    aria-label="!1:​example.org"
    aria-selected="false"
    class="mx_AccessibleButton mx_RoomTile"
    role="treeitem"
    tabindex="-1"
  >
    <div
      class="mx_DecoratedRoomAvatar"
    >
      <span
        class="mx_BaseAvatar"
        role="presentation"
      >
        <span
          aria-hidden="true"
          class="mx_BaseAvatar_initial"
          style="font-size: 20.8px; width: 32px; line-height: 32px;"
        >
          !
        </span>
        <img
          alt=""
          aria-hidden="true"
          class="mx_BaseAvatar_image"
          data-testid="avatar-img"
          loading="lazy"
          src="data:image/png;base64,00"
          style="width: 32px; height: 32px;"
        />
      </span>
    </div>
    <div
      class="mx_RoomTile_titleContainer"
    >
      <div
        class="mx_RoomTile_title mx_RoomTile_titleWithSubtitle"
        tabindex="-1"
        title="!1:​example.org"
      >
        <span
          dir="auto"
        >
          !1:​example.org
        </span>
      </div>
      <div
        class="mx_RoomTile_subtitle"
        id="mx_RoomTile_messagePreview_!1:example.org"
        title="test message"
      >
        <span
          class="mx_RoomTile_subtitle_text"
        >
          test message
        </span>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="mx_RoomTile_badgeContainer"
    />
    <div
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Room options"
      class="mx_AccessibleButton mx_RoomTile_menuButton"
      role="button"
      tabindex="0"
    />
    <div
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Notification options"
      class="mx_AccessibleButton mx_RoomTile_notificationsButton"
      role="button"
      tabindex="-1"
    />
  </div>
</DocumentFragment>
`;

exports[`RoomTile when message previews are enabled should render a room without a message as expected 1`] = `
<DocumentFragment>
  <div
    aria-describedby="mx_RoomTile_messagePreview_!1:example.org"
    aria-label="!1:​example.org"
    aria-selected="false"
    class="mx_AccessibleButton mx_RoomTile"
    role="treeitem"
    tabindex="-1"
  >
    <div
      class="mx_DecoratedRoomAvatar"
    >
      <span
        class="mx_BaseAvatar"
        role="presentation"
      >
        <span
          aria-hidden="true"
          class="mx_BaseAvatar_initial"
          style="font-size: 20.8px; width: 32px; line-height: 32px;"
        >
          !
        </span>
        <img
          alt=""
          aria-hidden="true"
          class="mx_BaseAvatar_image"
          data-testid="avatar-img"
          loading="lazy"
          src="data:image/png;base64,00"
          style="width: 32px; height: 32px;"
        />
      </span>
    </div>
    <div
      class="mx_RoomTile_titleContainer"
    >
      <div
        class="mx_RoomTile_title"
        tabindex="-1"
        title="!1:​example.org"
      >
        <span
          dir="auto"
        >
          !1:​example.org
        </span>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="mx_RoomTile_badgeContainer"
    />
    <div
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Room options"
      class="mx_AccessibleButton mx_RoomTile_menuButton"
      role="button"
      tabindex="0"
    />
    <div
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Notification options"
      class="mx_AccessibleButton mx_RoomTile_notificationsButton"
      role="button"
      tabindex="-1"
    />
  </div>
</DocumentFragment>
`;

exports[`RoomTile when message previews are not enabled should render the room 1`] = `
<div>
  <div
    aria-label="!1:​example.org"
    aria-selected="false"
    class="mx_AccessibleButton mx_RoomTile"
    role="treeitem"
    tabindex="-1"
  >
    <div
      class="mx_DecoratedRoomAvatar"
    >
      <span
        class="mx_BaseAvatar"
        role="presentation"
      >
        <span
          aria-hidden="true"
          class="mx_BaseAvatar_initial"
          style="font-size: 20.8px; width: 32px; line-height: 32px;"
        >
          !
        </span>
        <img
          alt=""
          aria-hidden="true"
          class="mx_BaseAvatar_image"
          data-testid="avatar-img"
          loading="lazy"
          src="data:image/png;base64,00"
          style="width: 32px; height: 32px;"
        />
      </span>
    </div>
    <div
      class="mx_RoomTile_titleContainer"
    >
      <div
        class="mx_RoomTile_title"
        tabindex="-1"
        title="!1:​example.org"
      >
        <span
          dir="auto"
        >
          !1:​example.org
        </span>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="mx_RoomTile_badgeContainer"
    />
    <div
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Room options"
      class="mx_AccessibleButton mx_RoomTile_menuButton"
      role="button"
      tabindex="0"
    />
    <div
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Notification options"
      class="mx_AccessibleButton mx_RoomTile_notificationsButton"
      role="button"
      tabindex="-1"
    />
  </div>
</div>
`;
